linked_list/linked_list.py

```diff
@@ -86,32 +151,72 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        current = self.head
+        prev = None
+
+        while current:
+            temp = current.next
+            current.next = prev
+            prev = current
+            current = temp
+
+        self.head = prev
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        outside_walker = self.head
+        mid_walker = self.head
+        count = 0
+
+        while outside_walker:
+            outside_walker = outside_walker.next
+            count += 1
+            if count % 2 == 0:
+                mid_walker = mid_walker.next

```

Comment:
Just a note, that you could have the outside walker move twice and the mid_walker move once with each iteration.  Then you don't have to do this work with modulus.